### ISSUE TITLE
CPU microcode is currently not installed, this should be done for every installation for security and stability.

### ISSUE THREAD

**Issue by dylanmtaylor** [archlinux/archinstall#177]
This isn't something that should be a user option, it should just detect the CPU vendor (intel/amd) and install it by default for literally any installation type, for security and stability reasons, particularly against speculative execution attacks.

https://wiki.archlinux.org/index.php/microcode

The packages are intel-ucode and amd-ucode, and as described in the Arch wiki, using systemd-boot as that is what this installer uses, it's just a matter of adding this line to the bootloader entry config file:

initrd  /cpu_manufacturer-ucode.img

```
The latest microcode cpu_manufacturer-ucode.img must be available at boot time in your EFI system partition (ESP). The ESP must be mounted as /boot in order to have the microcode updated every time amd-ucode or intel-ucode is updated. Otherwise, copy /boot/cpu_manufacturer-ucode.img to your ESP at every update of the microcode package.
```

**Comment by dylanmtaylor** [comment]
Sample config file with microcode loading:

```
➜  archinstall git:(master) cat /boot/loader/entries/arch.conf 
title Arch Linux
linux /vmlinuz-linux
initrd /intel-ucode.img
initrd /initramfs-linux.img
options rd.luks.name=1ea54b52-e005-4438-afd9-e690e690bec4=cryptlvm root=/dev/vg0/root resume=/dev/vg0/swap rd.luks.options=discard i915.fastboot=1 quiet splash loglevel=3 rd.udev.log_priority=3 vt.global_cursor_default=0 rw
```

This assumes that the absolute path to intel-ucode.img is `/boot/intel-ucode.img`.

**Comment by advaithm** [comment]
I added a function to check the cpu manufacturer in 94456bf8 it shouldn't be to use it to check and install and configure the appropriate systemd entry

**Comment by m1ten** [comment]
Print something like `Do you want to install, [ucode]? [Y/n]`? ucode being auto detected.

**Comment by advaithm** [comment]
Actually this may raise some issues with Chinese cpus since without some confirming what they report as we have no way to check for them

**Comment by Torxed** [comment]
> Actually this may raise some issues with Chinese cpus since without some confirming what they report we have no way to check for them

We can casually log an error if unsupported ucode is detected for now and skip it, and they can/should add it manually. We could also print a link to the wiki section for this if that happens.

**Comment by dylanmtaylor** [comment]
> Print something like `Do you want to install, [ucode]? [Y/n]`? ucode being auto detected.

Honestly, do you really want to ask for another choice? If we can't detect, a warning might be output but for amd/intel, not installing microcode can/will cause issues.

> I added a function to check the cpu manufacturer in [94456bf](https://github.com/archlinux/archinstall/commit/94456bf8ba692fb33c819d416636b34d9b14f7af) it shouldn't be to use it to check and install and configure the appropriate systemd entry

This function should probably be if/else if/else; right now it assumes that if it isn't AMD it's Intel which is technically wrong.

**Comment by dylanmtaylor** [comment]
> Actually this may raise some issues with Chinese cpus since without some confirming what they report we have no way to check for them

Does Arch even package ucode for these? I think only AMD/Intel are packaged.

**Comment by advaithm** [comment]
Arch doesn't ship ucodes for them but using amd/Intel ucodes could make the install non bootable

**Comment by dylanmtaylor** [comment]
> Arch doesn't ship ucodes for them but using amd/Intel ucodes could make the install non bootable

Yes, so that is why the flow would look like this:

* Check for AMD/Intel CPU
* If detected, based on the vendor, add microcode package and add to systemd-boot configuration
* Otherwise, output a warning and continue without adding ucode.

By doing this, we shouldn't break boot for these other vendors.

**Comment by Torxed** [comment]
> > Arch doesn't ship ucodes for them but using amd/Intel ucodes could make the install non bootable
> 
> Yes, so that is why the flow would look like this:
> 
> * Check for AMD/Intel CPU
> * If detected, based on the vendor, add microcode package and add to systemd-boot configuration
> * Otherwise, output a warning and continue without adding ucode.
> 
> By doing this, we shouldn't break boot for these other vendors.

Sounds reasonable and good. If no one else joinks it in the next couple of days I'll get right to it. Will probably be merged in `v2.2.0` tho for introducing new functionality that may or may not break things (it should improve things, but I have a sneaking suspicion something will go haywire)

**Comment by dylanmtaylor** [comment]
> > > Arch doesn't ship ucodes for them but using amd/Intel ucodes could make the install non bootable
> > 
> > 
> > Yes, so that is why the flow would look like this:
> > 
> > * Check for AMD/Intel CPU
> > * If detected, based on the vendor, add microcode package and add to systemd-boot configuration
> > * Otherwise, output a warning and continue without adding ucode.
> > 
> > By doing this, we shouldn't break boot for these other vendors.
> 
> Sounds reasonable and good. If no one else joinks it in the next couple of days I'll get right to it. Will probably be merged in `v2.2.0` tho for introducing new functionality that may or may not break things (it should improve things, but I have a sneaking suspicion something will go haywire). Unless it's urgent and actually causes issues, then I'll try to force it into the next patch.

I looked into this a bit and I think the change needs to be made here (adding a conditional on CPU manufacturer and formatting the entry file to include microcode loading)?

https://github.com/archlinux/archinstall/blob/master/archinstall/lib/installer.py#L365

Probably also need to modify the GRUB bootloader for BIOS systems as well, which is in the V2.2.0 branch?

**Comment by advaithm** [comment]
GRUB will auto detect and add it if its present @dylanmtaylor

**Comment by dylanmtaylor** [comment]
> GRUB will auto detect and add it if its present @dylanmtaylor

Ah, you're absolutely right. For some reason I saw that it needed to be added to the systemd-boot entry file and I saw the example of the GRUB config, showing the initrd line with the microcode explicitly listed. I guess I should have read the manual more closely.

```
grub-mkconfig will automatically detect the microcode update and configure GRUB appropriately. After installing the microcode package, regenerate the GRUB config to activate loading the microcode update by running:

# grub-mkconfig -o /boot/grub/grub.cfg
```

**Comment by Torxed** [comment]
> I looked into this a bit and I think the change needs to be made here (adding a conditional on CPU manufacturer and formatting the entry file to include microcode loading)?
> 
> https://github.com/archlinux/archinstall/blob/master/archinstall/lib/installer.py#L365

Correct, that's the place to do it! And we should probably place the hardware detection in [hardware.py](https://github.com/archlinux/archinstall/blob/master/archinstall/lib/hardware.py) :)

**Comment by PandaFoss** [comment]
I don't really know much about Python, but do you think it would be a performance improvement to replace the `lscpu` command with a line-by-line read of the `/proc/cpuinfo` file? This would mean looking for the string `GenuineIntel` or `AuthenticAMD` to identify an Intel or AMD CPU respectively.

**Comment by advaithm** [comment]
thats what my function does [here](https://github.com/archlinux/archinstall/blob/610d630863b949723431c230c4de8973dad8f60b/archinstall/lib/hardware.py#L8) it should be as easy as replacing AMD here

**Comment by Torxed** [comment]
> thats what my function does [here](https://github.com/archlinux/archinstall/blob/610d630863b949723431c230c4de8973dad8f60b/archinstall/lib/hardware.py#L8) it should be as easy as replacing AMD here

I was looking for that function, was pretty sure you added it a while back.
But I was looking for `getCPUVendor` or something, don't know why.

Perhaps `lscpu -J` could be of use here and just do:

```python
cpu_info = json.loads(b"".join(archinstall.sys_command('lscpu -J')).decode('UTF-8'))
```

**Comment by dylanmtaylor** [comment]
> I don't really know much about Python, but do you think it would be a performance improvement to replace the `lscpu` command with a line-by-line read of the `/proc/cpuinfo` file? This would mean looking for the string `GenuineIntel` or `AuthenticAMD` to identify an Intel or AMD CPU respectively.

Either one should execute so quickly (and is a one-time action), that it is likely negligible even on the slowest systems I'd imagine.

I decided to do a quick test reading /proc/cpuinfo with cat and running lscpu. 

```
lscpu  0.01s user 0.00s system 32% cpu 0.044 total
```

```
cat /proc/cpuinfo  0.00s user 0.00s system 14% cpu 0.025 total
```

So I guess just implement it however is easiest/cleanest, optimization isn't super important to save .01s.

**Comment by Torxed** [comment]
I agree, since this library (to the best of my knowledge) is run once. And nothing executes 10k+ times per second, any performance impact that isn't 1sec+ and re-occuring is negligible.

**Comment by grazzolini** [comment]
Getting microcode installed by default should be ok. However, it isn't recommended to install microcode on VM's, so, the installer should detect if it's on a VM and not install microcode.

**Comment by advaithm** [comment]
I can add a isVM function to check if we are running on a virtual machine

**Comment by grazzolini** [comment]
Let's use systemd-detect-virt for that.

**Comment by dylanmtaylor** [comment]
> Let's use systemd-detect-virt for that.

That's actually a really good idea. Seconded.

**Comment by advaithm** [comment]
It should be fairly easy to abstract it in hardware.py

**Comment by Torxed** [comment]
#438 fixed this together with #221.
